### PR TITLE
PR: Add --safe-mode option to start with a clean config directory

### LINKED ISSUE #7112: Running ./bootstap.py --test should open an empty history.py
### Issue Report Checklist

* [x] Searched the [issues page](https://github.com/spyder-ide/spyder/issues?q=is%3Aissue) for similar reports
* [x] Read the relevant sections of the [Spyder Troubleshooting Guide](https://github.com/spyder-ide/spyder/wiki/Troubleshooting-Guide-and-FAQ) and followed its advice
* [x] Reproduced the issue after updating with ``conda update spyder`` (or ``pip``, if not using Anaconda)
* [x] Could not reproduce inside ``jupyter qtconsole`` (if console-related)
* [x] Tried basic troubleshooting (if a bug/error)
    * [x] Restarted Spyder
    * [x] Reset preferences with ``spyder --reset``
    * [x] Reinstalled the latest version of [Anaconda](https://www.anaconda.com/download/)
    * [x] Tried the other applicable steps from the Troubleshooting Guide
* [x] Completed the **Problem Description**, **Steps to Reproduce** and **Version** sections below

## Problem Description
I found that ./bootstap.py --test would also open a default Spyder instance for debugging.

But it opens your local history and opens new file with my template

It would be easier to debug issues with out resetting my Syder config all the time.

### What steps reproduce the problem?

1. Start Spyder Master
2. See your 10+mb(I guess I like my code cells) history.py in the history
3. Start a new file and see your template

### What is the expected output? What do you see instead?

Like temporary a spyder reset 

## Versions

From anaconda in a new environ.
Spyder started from the command line with ```./bootstrap.py --test```

* Spyder version: current master
* Python version: 3.6.5 64bits
* Qt version: 5.9.4
* PyQt version: 5.9.2
* Operating System name/version: Ubuntu 16.04

### Dependencies

```
pyflakes >=0.6.0 :  1.6.0 (OK)
pycodestyle >=2.3:  2.4.0 (OK)
pygments >=2.0   :  2.2.0 (OK)
pandas >=0.13.1  :  None (NOK)
numpy >=1.7      :  None (NOK)
sphinx >=0.6.6   :  1.7.4 (OK)
rope >=0.9.4     :  0.10.7 (OK)
jedi >=0.11.0    :  0.12.0 (OK)
psutil >=0.3     :  5.4.5 (OK)
nbconvert >=4.0  :  5.3.1 (OK)
sympy >=0.7.3    :  None (NOK)
cython >=0.21    :  None (NOK)
qtconsole >=4.2.0:  4.3.1 (OK)
IPython >=4.0    :  6.4.0 (OK)
pylint >=0.25    :  1.8.4 (OK)
```

### PR DESCRIPTION
**Pull request by CAM-Gerlach**
### Pull Request Checklist

* [x] Read and followed this repo's [Contributing Guidelines](https://github.com/spyder-ide/spyder/blob/master/CONTRIBUTING.md)
* [x] Based your PR on the latest version of the correct branch (master or 3.x)
* [x] Followed [PEP8](https://www.python.org/dev/peps/pep-0008/) for code style
* [x] Ensured your pull request hasn't eliminated unrelated blank lines/spaces,
      modified the ``spyder/defaults`` directory, or added new icons/assets
* [x] ~~Wrote at least one-line docstrings for any new functions~~
* [x] ~~Added at least one unit test covering the changes, if at all possible~~
* [x] Described your changes and the motivation for them below
* [x] Noted what issue(s) this pull request resolves, creating one if needed
* [x] ~~Included a screenshot, if this PR makes any visible changes to the UI~~

## Description of Changes

Thanks to @bcolsen for his detailed report leading to this PR.

Currently, passing the ``--test`` flag to ``bootstrap.py`` just sets the ``SPYDER_TEST`` env variable which instead tells spyder to not load and save the user's individual settings file, e.g. ``spyder.ini``, while not affecting handling of anything else inside for ``spyder-py3`` directory, including the history logs for the normal and internal consoles, along with your pylint, profiler, working dir, help, and online help histories, your language and ``rope`` settings, any third party plugins you may have installed, and your default ``temp.py`` and ``template.py`` (new file template) files. Obviously, this is not particularly desirable for most testing scenarios—one would want "clean" to be truly "clean", without such cruft. 

Therefore, this PR just makes ``--test`` set the Spyder config dir to a clean temp one, just like ``SPYDER_PYTEST`` currently does. It also makes sure to clean the current temp directory if one exists, removes the old, unnecessary code to just load a clean config file, and does some minor cleanup immediately around changed lines.

It seems its not really possible to include a unit test for this as the Pytest suite already does this anyway, and has other checks for it, if its otherwise even really possible to test for.

### Issue(s) Resolved

Fixes #7112

### CODE REVIEW

**Comment by CAM-Gerlach** [comment]
Wrt @pep8speaks , I was just following the convention already in the file for all other spyder imports, it is likely needed for safety, and I added # analysis:ignore for a good measure.

**Comment by bcolsen** [comment]
I tested it on my home computer and it seem to work well.  

I'll test it a work tomorrow against the 10 mb history.py :-)

**Comment by CAM-Gerlach** [comment]
> I'll test it a work tomorrow against the 10 mb history.py :-)

I'm not sure I understand how the size of any ``history.py`` file in the normal Spyder config dir would have any effect on ``--test`` mode whatsoever. Either it works (using the temp config dir), or it doesn't (it still uses the main one), which would be obvious already on your home machine assuming you had anything at all in the history before.

**Comment by bcolsen** [comment]
Sorry bad joke! You are right the file size would clearly have no affect.

I just had another idea.  We could add this to the main Spyder executable as well to make trouble shooting easier.  This would be like having a "safe mode" like Firefox has. It's like testing a Spyder reset, but if it doesn't fix the problem you don't lose your settings. We could even call it --safe-mode.

**Comment by CAM-Gerlach** [comment]
> Sorry bad joke! You are right the file size would clearly have no affect.

I've certainly had more than my share of those, so I'm not one to talk...

> I just had another idea. We could add this to the main Spyder executable as well to make trouble shooting easier.

Great idea! Shouldn't be hard to do; the only thing is that normal Spyder and ``bootstrap.py`` use two separate argument parsers, so we'd have to either stick it in both (very easy to do, but not exactly DRY) or refactor that, and move the small block of code that wipes the temp config to near the beginning of ``start.py``. @ccordoba12 , agree?

**Comment by CAM-Gerlach** [comment]
The broader change I would suggest is *unifying the bootstrap parser into the main ``cli_options`` ``get_options()`` parser since the great majority if not all of their options are useful to either start mode*; in particular, this relates to your #7123 , since you can't actually set ``spyder --debug`` with normal Spyder even though that's what we've been telling a lot of users to help figure out problems (we've instead had to have them manually set the `SPYDER_DEBUG` environment variable to `3`, in a platform unportable way). 

*If we needed a few different options for each, we could either add an parameter to ``get_options()`` in ``cli_options`` to tell it which set of extra options to add, or we could just add them to the parser returned by ``get_options()`` in ``start.py`` and ``boostrap.py`` respectively.* Then, the only other thing we'd need to change is to have start.main() take one parameter, ``options`` or the like, with a default value of None that bootstrap.py could pass the rest of the parsed options that start.py would need, and then enclose the current parser block in ``start.py`` with ``if options is None:`` which would replicate the current behavior by default.

However, the one complexity here is that Spyder 4 was converted over to use ``argparse``, while Spyder 3 is still stuck on the deprecated ``argparse``, which will make more complicated, at least for @ccordoba12 when merging it to ``master``. We could do this just in ``master``, but since its not really a feature but rather an important change helping users find and solve bugs and problems, and would make our lives a good deal easier maintaining Spyder 3, I would think we would want to get it out to them as soon as possible. *So, we can either keep things using the deprecated optparse in Spyder 3 and Carlos can fix any conflicts merging to master, or as part of this we can backport the argparse change.*

@ccordoba12 , opinions? I italicized the two portions most in need of your input so you could find them easily.

File: spyder/config/base.py [hunk 1]
```diff
@@ -144,15 +144,15 @@ def get_conf_path(filename=None):
...
     # Create conf_dir
     if not osp.isdir(conf_dir):
-        if running_under_pytest():
+        if running_under_pytest() or TEST:
             os.makedirs(conf_dir)
         else:
             os.mkdir(conf_dir)
     if filename is None:
         return conf_dir
-    else:
```
Comment (ccordoba12) [inline spyder/config/base.py]: This needs to stay, so please revert this change. It's how we retrieve files placed placed under our config directory, like this `get_conf_path('history.py')`

**Comment by ccordoba12** [comment]
> This would be like having a "safe mode" like Firefox has. It's like testing a Spyder reset, but if it doesn't fix the problem you don't lose your settings. We could even call it --safe-mode.

We're using pytest for this, so this we'll need to improve our current implementation.

> The broader change I would suggest is unifying the bootstrap parser into the main cli_options get_options() parser since the great majority if not all of their options are useful to either start mode

I don't know about this. I'll have a PR with this change, but the initial idea behind different cli options for both is that boostrap options don't apply to the main Spyder script and vice versa.

**Comment by ccordoba12** [comment]
> I didn't mean to have this setting for unit testing but as a convenient way for users with bugs to troubleshoot their local install problems without deleting all their settings

I understood that. So, if we want to add an option to use a clean prefs directory as part of a regular Spyder installation, we'd need to not use pytest for the implementation because then pytest would become a Spyder dependency, and we don't want that.

**Comment by bcolsen** [comment]
Sorry for my confusion. I get it now, thanks.  It will require more work than I thought.

File: spyder/config/base.py (hunk 1 above)
Comment (CAM-Gerlach) [reply spyder/config/base.py]: I can, but there is no functional difference to dropping the ``else,`` it works exactly the same (or I would have noticed immediately since this was exactly the functionality I was testing)—did you scroll down a few lines? if filename is not None, it just executes the next statement, which is ``return osp.join(conf_dir, filename)`` exactly as before. I just cleaned it up while I was at it since it was in the same function I made other more substantial changes to.

**Comment by CAM-Gerlach** [comment]
> So, if we want to add an option to use a clean prefs directory as part of a regular Spyder installation, we'd need to not use pytest for the implementation because then pytest would become a Spyder dependency, and we don't want that.

Good point, I didn't think of that. In fact, since pytest (and py, which is also needed) aren't dependencies for a basic source install either, even in its current form its a problem. 

However, upon further investigation, it actually would be quite simple (and both in scope and necessary for just this PR even, given the above) as well as probably optimal overall to just reimplement the two relevant functions (one from ``pytest`` and one from ``py``), because both packages are imported locally just for those two functions and used nowhere else in the file, and both functions are just one-line thin wrappers around standard library functions;

* ``py.path.local.get_temproot()`` is a oneliner that just ``import``s ``tempfile``,  calls ``tempfile.gettempdir()`` and then casts it to a py.path.local convenience wrapper, which in our code is then immediately casted back to ``str`` making it pointless and wasteful to use the wrapper instead of just ``tempfile.gettempdir()`` directly.
* ``_pytest.tmpdir.get_user`` is a private (so should we even be relying on it?) wrapper for ``getpass.getuser()`` that just calls it in a try-except block, returning ``None`` if it doesn't work due to a narrow set of conditions (Windows + a ``tox`` based CI that strips env variables). It would be trivial to include just try-except in ``base.py``, if it is even really necessary in our use case, or even including the user name to begin with—on Windows, the one platform where the standard library get_user() could fail, each user has their own temp directory such that including the username is unnecessary anyway

Therefore, I'll implement those two minor changes here, and we can go from there.

>  I'll have to see a PR with this change, but the initial idea behind different cli options for both is that boostrap options don't necessarily apply to the main Spyder script and vice versa.

I'll submit a separate issue and PR. Looking through the existing CLI options for normal Spyder, it looks like pretty much every single one of them would be applicable to ``boostrap.py`` as far as I can tell; on the other hand, two of the six in bootstrap.py's are identical to those in ``cli_options``, (show and hide console), two would be very useful to have there (``test`` and ``debug``), and two are ``bootstrap`` specific (``no-apport`` and ``gui``). 

Therefore, would seem the best course of action might be to add the two commonly useful ones to ``cli_options``, do a small refactor to get_options so it returns the parser rather than the evaluated options (its only one extra line, and IIRC a best practice anyway), and then add the two extra options in ``boostrap.py``, passing the evaluated options on to main(). All in all, maybe a dozen lines worth of changes.

**Comment by CAM-Gerlach** [comment]
Okay, I implemented what I discussed in the first part of previous post to remove the dependencies on ``py`` and ``pytest`` in the relevant portions; overall its actually a small net improvement to do so even aside from the proximate reason. I tested it with and without ``clean``, doing various activities in Spyder and restarting several times, and everything related to the change works well. Therefore, this PR looks ready on my end.

With regard to the second point, I figured out it already works to call ``bootstrap.py --clean -- --spyder-options-here`` so there seems little point to doing most of what I suggest above. However, this just means we can move ``debug`` and ``clean`` over to the main parser, handle them appropriately in ``start.py`` and/or ``mainwindow`` instead of ``boostrap.py``, and be done with it—I'll just need to make sure they "stick" before they're needed. I'll open another issue/PR to take a crack at that.

**Comment by bcolsen** [comment]
I tested this and it works on linux :+1:

File: spyder/config/base.py (hunk 1 above)
Comment (ccordoba12) [reply spyder/config/base.py]: This is a critical part of our settings infrastructure, so I don't want to play with it simply for correctness, so please revert this change.

Comment (CAM-Gerlach) [reply spyder/config/base.py]: Okay, done as requested. 

In general, I abstain from making these sorts of changes (that aren't purely code style) for the reasons you allude to, namely the small but non-negligible chance of breaking something that's rarely worth the lesser benefit from cleaner code. In this case I ended up doing so since I'd checked it with both my own knowledge and the static analyzer, and the fact that the unit tests and numerous real world-tests exercising that block of code all had no issues, but it obviously wasn't worth doing and I'll be even more conservative in the future.

**Comment by CAM-Gerlach** [comment]
@ccordoba12 Addressed your comment; thanks.

File: bootstrap.py [hunk 2]
```diff
@@ -39,8 +42,8 @@
                   "is to show the console")
 parser.add_option('--hide-console', action='store_true',
                   default=False, help="Hide parent console window (Windows only)")
-parser.add_option('--test', dest="test", action='store_true', default=False,
-                  help="Test Spyder with a clean settings dir")
+parser.add_option('--clean', dest="clean", action='store_true', default=False,
```
Comment (ccordoba12) [inline bootstrap.py]: Please change this to `--clean-config`. The problem is `--clean` is too close in meaning to `--reset`.

File: bootstrap.py [hunk 3]
```diff
@@ -54,9 +57,9 @@
 assert options.gui in (None, 'pyqt5', 'pyqt', 'pyside'), \
        "Invalid GUI toolkit option '%s'" % options.gui
 
-# For testing purposes
-if options.test:
-    os.environ['SPYDER_TEST'] = 'True'
+# Start Spyder with a clean configuration directory for testing purposes
+if options.clean:
```
Comment (ccordoba12) [inline bootstrap.py]: This should be `options.clean_config`

File: bootstrap.py [hunk 4]
```diff
@@ -54,9 +57,9 @@
...
+    os.environ['SPYDER_CLEAN'] = 'True'
```
Comment (ccordoba12) [inline bootstrap.py]: `SPYDER_CLEAN` -> `SPYDER_CLEAN_CONFIG`

File: spyder/config/base.py [hunk 5]
```diff
@@ -34,9 +34,9 @@
 # SPYDER_DEV is (and *only* has to be) set in bootstrap.py
 DEV = os.environ.get('SPYDER_DEV')
 
-# For testing purposes
-# SPYDER_TEST can be set using the --test option of bootstrap.py
-TEST = os.environ.get('SPYDER_TEST')
+# Make Spyder use a temp clean configuration directory for testing purposes
+# SPYDER_CLEAN can be set using the --clean option of bootstrap.py
+CLEAN = os.environ.get('SPYDER_CLEAN')
```
Comment (ccordoba12) [inline spyder/config/base.py]: This should be

    CLEAN_CONFIG = os.environ.get('SPYDER_CLEAN_CONFIG')

File: spyder/config/base.py [hunk 6]
```diff
@@ -122,13 +122,18 @@ def get_home_dir():
...
+    if running_under_pytest() or CLEAN:
+        # Use clean config dir if running tests or the user requests it.
+        import getpass  # analysis:ignore
+        if sys.platform.startswith("win"):
+            current_user = ''
+        else:
+            current_user = '-' + str(getpass.getuser())
+        import tempfile  # analysis:ignore
+        conf_dir = osp.join(str(tempfile.gettempdir()),
+                            'pytest-spyder{0!s}'.format(current_user),
```
Comment (ccordoba12) [inline spyder/config/base.py]: Let's move this whole block to a new function called `clean_conf_dir`

File: spyder/config/base.py [hunk 7]
```diff
@@ -144,15 +149,15 @@ def get_conf_path(filename=None):
...
-        if running_under_pytest():
+        if running_under_pytest() or CLEAN:
             os.makedirs(conf_dir)
         else:
             os.mkdir(conf_dir)
     if filename is None:
         return conf_dir
     else:
         return osp.join(conf_dir, filename)
-
```
Comment (ccordoba12) [inline spyder/config/base.py]: Please avoid adding/removing blanks in the 3.x branch. It makes my life harder when merging with `master` and also with `split-plugins`. All pep8 stuff needs to be done in `master`, not here, because this branch is in maintenance now.

File: spyder/config/main.py [hunk 8]
```diff
@@ -12,21 +12,21 @@
...
 from spyder.config.user import UserConfig
 from spyder.config.utils import IMPORT_EXT
 from spyder.utils import codeanalysis
 
 
-#==============================================================================
+# =============================================================================
 # Main constants
-#==============================================================================
+# =============================================================================
```
Comment (ccordoba12) [inline spyder/config/main.py]: The same goes for this. No worries in this PR, but please avoid it in the future.

File: bootstrap.py (hunk 2 above)
Comment (CAM-Gerlach) [reply bootstrap.py]: Good point @ccordoba12 . On that note, ``--clean-config`` could also be misinterpreted in the same way, i.e. that the option would ``clean`` the ``config`` i.e. what ``reset`` does. What about ``--clean-mode``? That would make it much clearer that its temporary and distinct from ``reset``.

**Comment by ccordoba12** [comment]
I left several new comments because I don't like `--clean` as an option.

File: spyder/config/base.py (hunk 7 above)
Comment (CAM-Gerlach) [reply spyder/config/base.py]: Sorry; my mistake. It must have happened when I did the ``else:`` change; one more reason to avoid that sort of thing I suppose.

File: spyder/config/main.py (hunk 8 above)
Comment (CAM-Gerlach) [reply spyder/config/main.py]: Okay, noted and will now do so from here on out.

File: bootstrap.py (hunk 2 above)
Comment (ccordoba12) [reply bootstrap.py]: Firefox call it `--safe-mode`. What about that?

Comment (ccordoba12) [reply bootstrap.py]: You're right about `--clean-config`, by the way.

File: spyder/config/base.py (hunk 6 above)
Comment (CAM-Gerlach) [reply spyder/config/base.py]: Will do, thanks.

File: bootstrap.py (hunk 2 above)
Comment (CAM-Gerlach) [reply bootstrap.py]: Done.

File: bootstrap.py (hunk 3 above)
Comment (CAM-Gerlach) [reply bootstrap.py]: Changed to ``options.safe_mode``.

File: bootstrap.py (hunk 4 above)
Comment (CAM-Gerlach) [reply bootstrap.py]: Changed to ``SPYDER_SAFE_MODE``.

File: spyder/config/base.py (hunk 5 above)
Comment (CAM-Gerlach) [reply spyder/config/base.py]: Changed to ``SAFE_MODE``/``SPYDER_SAFE_MODE`` respectively.

File: spyder/config/base.py (hunk 6 above)
Comment (CAM-Gerlach) [reply spyder/config/base.py]: Done; moved to ``get_clean_config_dir``.

**Comment by ccordoba12** [comment]
Ok, no problem. Good catch!
